pulp_rpm/app/modelresource.py

```diff
@@ -346,10 +353,145 @@ def set_up_queryset(self):
 
     class Meta:
         model = RpmRepository
-        import_id_fields = ('name',)
+        import_id_fields = ("name",)
         fields = ("name", "pulp_type", "description", "original_checksum_types", "sub_repo")
 
 
+class UpdateCollectionResource(QueryModelResource):
+    """
+    Resource for import/export of rpm_updatecollection entities.
+    """
+
+    update_record = fields.Field(
+        column_name="update_record",
+        attribute="update_record",
+        widget=ForeignKeyWidget(UpdateRecord, field="digest"),
+    )
+
+    def set_up_queryset(self):
+        """
+        Set up a queryset for UpdateCollections.
+
+        Returns:
+            UpdateCollections belonging to UpdateRecords for a specified repo-version.
+
+        """
+        return UpdateCollection.objects.filter(
+            update_record__in=UpdateRecord.objects.filter(
+                pk__in=self.repo_version.content
+            )
+        )
+
+    class Meta:
+        model = UpdateCollection
+        exclude = QueryModelResource.Meta.exclude + ("packages",)
+        import_id_fields = tuple(
+            chain.from_iterable(UpdateCollection._meta.unique_together)
+        )
+
+
+class UpdateReferenceResource(QueryModelResource):
+    """
+    Resource for import/export of rpm_updatereference entities.
+    """
+
+    update_record = fields.Field(
+        column_name="update_record",
+        attribute="update_record",
+        widget=ForeignKeyWidget(UpdateRecord, field="digest"),
+    )
+
+    def set_up_queryset(self):
+        """
+        Set up a queryset for UpdateReferences.
+
+        Returns:
+            UpdateReferences belonging to UpdateRecords for a specified repo-version.
+
+        """
+        return UpdateReference.objects.filter(
+            update_record__in=UpdateRecord.objects.filter(pk__in=self.repo_version.content)
+        )
+
+    class Meta:
+        model = UpdateReference
+        import_id_fields = (
+            "href",
+            "ref_type",
+            "update_record",
+        )
+
+
+class UpdateCollectionPackageResource(QueryModelResource):
+    """
+    Resource for import/export of rpm_updatecollectionpackage entities.
+    """
+
+    class UpdateCollectionForeignKeyWidget(ForeignKeyWidget):
+        """
+        Class that lets us specify a multi-key link to UpdateCollection.
+
+        Format is str(<name>|<update_record.digest>), to be used at import-row time.
+        """
+
+        def render(self, value, obj):
+            """Render formatted string to use as unique-identifier."""
+            return f"{obj.update_collection.name}|{obj.update_collection.update_record.digest}"
+
+    update_collection = fields.Field(
+        column_name="update_collection",
+        attribute="update_collection",
+        widget=UpdateCollectionForeignKeyWidget(UpdateCollection),
+    )
+
+    def before_import_row(self, row, **kwargs):
```

Comment:
Does this run before each individual object being imported or is it run once?

Reply:
At import-time, this is called before each row (ie per-object/instance)

Reply:
How is the performance of this?  This looks like 2 many-to-many lookups per N UpdateCollectionPackages... and there are a lot of UpdateCollectionPackages.

Reply:
Good question - altho you're only going back to the database when the UpdateRecord/UpdateCollection in question aren't in django's cache., and there is a LOT of 'locality of reference' going on here. There are possibly/probably ways to improve performance here - can you open a Story to measure and possibly improve?

Reply:
The Django cache is pretty limited in what it can actually do, I'm 99% certain it wouldn't touch the cache.  Caching only really helps if you're iterating the same queryset more than once.

Agree, let's not be specific though and just call it a general investigation of import/export performance.

Reply:
Sounds like a plan!